File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/client/resources/GameResources.java
package pize.tests.voxelgame.client.resources;

import pize.app.Disposable;
import pize.audio.sound.AudioBuffer;
import pize.audio.sound.AudioLoader;
import pize.files.Resource;
import pize.graphics.texture.Pixmap;
import pize.graphics.texture.PixmapIO;
import pize.graphics.texture.Region;
import pize.graphics.texture.Texture;
import pize.graphics.texture.atlas.TextureAtlas;
import pize.tests.voxelgame.main.Identifier;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GameResources implements Disposable{

    private final TextureAtlas<String> blockAtlas;
    private final List<Resource> blocksToLoadList;

    private final Map<String, AudioBuffer> audioList;

    public GameResources(){
        blockAtlas = new TextureAtlas<>();
        blocksToLoadList = new ArrayList<>();

        audioList = new HashMap<>();
    }


    public void load(){
        System.out.println("[Client]: Build blocks atlas");

        for(Resource resource: blocksToLoadList){
            final String name = resource.getSimpleName();
            final Pixmap pixmap = PixmapIO.load(resource);
            blockAtlas.put(name, pixmap);
        }

        blockAtlas.generate(512, 512);
    }


    public void registerBlockRegion(Resource resource){
        blocksToLoadList.add(resource);
    }

    public void registerBlockRegion(String path, String name){
        registerBlockRegion(new Resource(path + name + ".png"));
    }

    public Region getBlockRegion(String name){
        return blockAtlas.getRegion(name);
    }

    public Texture getBlocks(){
        return blockAtlas.getTexture();
    }


    public void registerSound(String soundID, Resource resource){
        System.out.println("Load sound " + soundID);
        final AudioBuffer buffer = new AudioBuffer();
        AudioLoader.load(buffer, resource);
        audioList.put(soundID, buffer);
    }

    public void registerSound(String soundID, String path, String name){
        registerSound(Identifier.namespaceID(soundID), new Resource(path + name + ".ogg"));
    }

    public void registerSound(String path, String name){
        final String soundID = name.replace("/", ".");
        registerSound(soundID, path, name);
    }

    public void registerSound(String path, String name, int min, int max){
        for(int i = min; i <= max; i++){
            final String soundID = name.replace("/", ".") + "." + i;
            registerSound(soundID, path, name + i);
        }
    }

    public AudioBuffer getSound(String soundID){
        return audioList.get(soundID);
    }


    @Override
    public void dispose(){
        blocksToLoadList.clear();
        blockAtlas.getTexture().dispose();
        for(AudioBuffer audio: audioList.values())
            audio.dispose();
    }

}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/client/resources/GameResources.kt
package pize.tests.voxelgame.client.resources

import pize.app.Disposable
import pize.audio.sound.AudioBuffer
import pize.audio.sound.AudioLoader.load
import pize.files.Resource
import pize.graphics.texture.PixmapIO.load
import pize.graphics.texture.Region
import pize.graphics.texture.Texture
import pize.graphics.texture.atlas.TextureAtlas
import pize.tests.voxelgame.main.Identifier

class GameResources : Disposable {
    private val blockAtlas: TextureAtlas<String>
    private val blocksToLoadList: MutableList<Resource>
    private val audioList: MutableMap<String?, AudioBuffer>

    init {
        blockAtlas = TextureAtlas()
        blocksToLoadList = ArrayList()
        audioList = HashMap()
    }

    fun load() {
        println("[Client]: Build blocks atlas")
        for (resource in blocksToLoadList) {
            val name = resource.simpleName
            val pixmap = load(resource)
            blockAtlas.put(name, pixmap)
        }
        blockAtlas.generate(512, 512)
    }

    fun registerBlockRegion(resource: Resource) {
        blocksToLoadList.add(resource)
    }

    fun registerBlockRegion(path: String, name: String) {
        registerBlockRegion(Resource("$path$name.png"))
    }

    fun getBlockRegion(name: String): Region? {
        return blockAtlas.getRegion(name)
    }

    val blocks: Texture?
        get() = blockAtlas.texture

    fun registerSound(soundID: String?, resource: Resource?) {
        println("Load sound $soundID")
        val buffer = AudioBuffer()
        load(buffer, resource!!)
        audioList[soundID] = buffer
    }

    fun registerSound(soundID: String, path: String, name: String) {
        registerSound(Identifier.namespaceID(soundID), Resource("$path$name.ogg"))
    }

    fun registerSound(path: String, name: String) {
        val soundID = name.replace("/", ".")
        registerSound(soundID, path, name)
    }

    fun registerSound(path: String, name: String, min: Int, max: Int) {
        for (i in min..max) {
            val soundID = name.replace("/", ".") + "." + i
            registerSound(soundID, path, name + i)
        }
    }

    fun getSound(soundID: String?): AudioBuffer? {
        return audioList[soundID]
    }

    override fun dispose() {
        blocksToLoadList.clear()
        blockAtlas.texture!!.dispose()
        for (audio in audioList.values) audio.dispose()
    }
}
